core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -811,17 +795,30 @@ object IO {
         } yield fiberA.zipWith(fiberAs)(_ :: _)
     }
 
+  final def bracket0[E, A, B](
+    acquire: IO[E, A]
+  )(release: (A, ExitResult[E, B]) => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
+    IORef[E, Option[(A, ExitResult[E, B])]](None).flatMap { m =>
+      (for {
+        a <- acquire.uninterruptibly
```

Comment:
You need to set the ref in here.

The reason is that `use` can be interrupted before it starts, so the ref won't be set, which means the release won't happen.

This should be changed to something like:

```scala
a <- acquire.flatMap(a => m.write(Some((a, ExitResult.Terminated(???))) *> IO.now(a))).uninterruptibly
```

It also occurs to me that `run` is not interruptible so this version of bracket is not interruptible in `use`. 😱 

That problem can be fixed separately, however. I'm not looking forward to it. 😆